File: tools/memgate_test.py
#!/usr/bin/env python3
"""memgate_test.py — unit tests for the memory-pressure gate (pure parsing logic)."""
import importlib.util
import sys
import os

spec = importlib.util.spec_from_file_location(
    "memgate", os.path.join(os.path.dirname(__file__), "memgate.py")
)
memgate = importlib.util.module_from_spec(spec)
spec.loader.exec_module(memgate)


VM_STAT_SAMPLE = """Mach Virtual Memory Statistics: (page size of 16384 bytes)
Pages free:                          1971366.
Pages active:                         286640.
Pages inactive:                       176020.
Pages speculative:                      1200.
Pages wired down:                     880000.
Pages purgeable:                        4047.
Pages occupied by compressor:          12345.
"""


def test_darwin_parsing(monkeypatch_cap=False):
    def fake_run(cmd):
        if cmd == ["vm_stat"]:
            return VM_STAT_SAMPLE
        if cmd == ["sysctl", "-n", "hw.pagesize"]:
            return "16384\n"
        if cmd == ["sysctl", "-n", "hw.memsize"]:
            return "38654705664\n"
        return ""

    orig = memgate._run
    memgate._run = fake_run
    try:
        d = memgate.read_darwin()
    finally:
        memgate._run = orig

    assert d["total_bytes"] == 38654705664, d
    # 1971366 free pages * 16384 = 32,314,533,888 bytes ~ 32.3GB
    assert d["free_bytes"] == 1971366 * 16384, d
    # 880000 wired * 16384 = 14,417,920,000 ~ 14.4GB (a resident 27B GPU model)
    assert d["wired_bytes"] == 880000 * 16384, d
    # available folds purgeable and keeps the 2GB safety margin
    expected_avail = d["free_bytes"] + 4047 * 16384 - int(memgate.SAFETY_MARGIN_GB * 1e9)
    assert d["available_bytes"] == max(expected_avail, 0), d
    print("test_darwin_parsing OK")


def test_snapshot_high_wired_flag():
    """When wired > 40% of RAM the snapshot flags a likely GPUresident."""
    orig_read = memgate.read_memory
    memgate.read_memory = lambda: {
        "total_bytes": 38654705664,
        "free_bytes": int(32e9),
        "available_bytes": int(30e9),
        "purgeable_bytes": int(1e9),
        "wired_bytes": int(16e9),  # 16GB / 38.6GB ~ 41% -> HIGH
        "compressed_bytes": 0,
    }
    orig_holders = memgate.big_holders
    memgate.big_holders = lambda: []
    try:
        snap = memgate.snapshot()
    finally:
        memgate.read_memory = orig_read
        memgate.big_holders = orig_holders
    assert snap["high_wired"] is True, snap
    assert "GPU" in snap["note"].upper() or "METAL" in snap["note"].upper(), snap["note"]
    print("test_snapshot_high_wired_flag OK")


def test_admit_decision_threshold():
    """admit is True iff available >= require AND wired is not high."""
    orig_read = memgate.read_memory
    orig_holders = memgate.big_holders

    def mem_with(free_gb, wired_gb):
        return {
            "total_bytes": 38654705664,
            "free_bytes": int(free_gb * 1e9),
            "available_bytes": int(free_gb * 1e9),
            "purgeable_bytes": 0,
            "wired_bytes": int(wired_gb * 1e9),
            "compressed_bytes": 0,
        }

    memgate.big_holders = lambda: []
    try:
        # Plenty free, low wired -> admit
        memgate.read_memory = lambda: mem_with(30, 5)
        snap = memgate.snapshot()
        assert snap["available_gb"] >= 20 and not snap["high_wired"]
        # Tight free -> would refuse a 20GB require
        memgate.read_memory = lambda: mem_with(10, 5)
        snap = memgate.snapshot()
        assert snap["available_gb"] < 20
        # High wired (GPUresident) -> flag even if 'free' looks ok
        memgate.read_memory = lambda: mem_with(25, 20)
        snap = memgate.snapshot()
        assert snap["high_wired"] is True
    finally:
        memgate.read_memory = orig_read
        memgate.big_holders = orig_holders
    print("test_admit_decision_threshold OK")


if __name__ == "__main__":
    failures = 0
    for t in [test_darwin_parsing, test_snapshot_high_wired_flag, test_admit_decision_threshold]:
        try:
            t()
        except AssertionError as e:
            print(f"FAIL {t.__name__}: {e}", file=sys.stderr)
            failures += 1
    print(f"\n{'OK' if not failures else str(failures)+' FAILED'} — memgate_test")
    sys.exit(1 if failures else 0)

File: internal/memgate/memgate_test.go
package memgate

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestParseLinux(t *testing.T) {
	mem := ParseLinux("MemTotal: 36000000 kB\nMemFree: 4000000 kB\nMemAvailable: 28000000 kB\nCached: 12000000 kB\n")
	if mem.TotalBytes != 36000000*1024 || mem.FreeBytes != 4000000*1024 {
		t.Fatalf("mem=%+v", mem)
	}
	if mem.AvailableBytes <= 0 || mem.PurgeableBytes != 12000000*1024 {
		t.Fatalf("mem=%+v", mem)
	}
}

func TestParseDarwin(t *testing.T) {
	vm := "Pages free: 1000.\nPages purgeable: 500.\nPages wired down: 2000.\nPages occupied by compressor: 10.\n"
	mem := ParseDarwin(vm, 4096, 16_000_000)
	if mem.FreeBytes != 1000*4096 || mem.PurgeableBytes != 500*4096 || mem.WiredBytes != 2000*4096 || mem.CompressedBytes != 10*4096 {
		t.Fatalf("mem=%+v", mem)
	}
}

// TestParseDarwinCountsInactiveAsAvailable pins the steady-state Mac case where the
// standalone `fak serve --gguf` Metal admission refused every load (issue #10595):
// macOS parks most reclaimable RAM in inactive (and speculative) pages, so a
// free+purgeable-only available figure reads ~0 on a healthy 36G host and the
// fail-closed reservation always refuses. The sample is the live vm_stat captured
// at the refusal: free 0.2G, inactive 16.6G, speculative ~0, purgeable 0.2G.
func TestParseDarwinCountsInactiveAsAvailable(t *testing.T) {
	vm := strings.Join([]string{
		"Mach Virtual Memory Statistics: (page size of 16384 bytes)",
		"Pages free:                                    10279.",
		"Pages active:                                1012697.",
		"Pages inactive:                              1012324.",
		"Pages speculative:                              1215.",
		"Pages throttled:                                   0.",
		"Pages wired down:                             154812.",
		"Pages purgeable:                               13107.",
	}, "\n")
	mem := ParseDarwin(vm, 16384, 38_654_705_664)
	reclaimable := int64(10279+1012324+1215+13107) * 16384
	if mem.AvailableBytes != reclaimable-int64(SafetyMarginGB*1e9) {
		t.Fatalf("AvailableBytes=%d, want inactive+speculative counted: %d", mem.AvailableBytes, reclaimable-int64(SafetyMarginGB*1e9))
	}
}

func TestParseHolders(t *testing.T) {
	holders := ParseHolders("PID RSS COMM\n123 2500000 llama-server\n456 100 shell\n789 1500000 python worker\n")
	if len(holders) != 2 || holders[0].PID != 123 || holders[1].PID != 789 {
		t.Fatalf("holders=%+v", holders)
	}
}

func TestBuildSnapshotAndEvaluate(t *testing.T) {
	mem := Memory{TotalBytes: 10_000_000_000, FreeBytes: 1_000_000_000, AvailableBytes: 5_000_000_000, WiredBytes: 5_000_000_000}
	s := BuildSnapshot("darwin", mem, nil)
	if !s.HighWired || s.Note == "ok" {
		t.Fatalf("snapshot=%+v", s)
	}
	e := Evaluate(s, 4)
	if e.Admit == nil || *e.Admit {
		t.Fatalf("high wired should refuse: %+v", e)
	}
	mem.WiredBytes = 0
	s = BuildSnapshot("linux", mem, nil)
	e = Evaluate(s, 4)
	if e.Admit == nil || !*e.Admit || e.ShortfallGB != 0 {
		t.Fatalf("expected admit: %+v", e)
	}
	e = Evaluate(s, 8)
	if e.Admit == nil || *e.Admit || e.ShortfallGB != 3 {
		t.Fatalf("expected shortfall: %+v", e)
	}
}

func TestAdmissionSampleForFailsClosedAndClassifiesPressure(t *testing.T) {
	tests := []struct {
		name string
		mem  Memory
		want Pressure
	}{
		{name: "unknown total", mem: Memory{AvailableBytes: 8 << 30}, want: PressureUnknown},
		{name: "unknown available", mem: Memory{TotalBytes: 16 << 30}, want: PressureUnknown},
		{name: "normal", mem: Memory{TotalBytes: 16 << 30, AvailableBytes: 8 << 30}, want: PressureNormal},
		{name: "warning compressor", mem: Memory{TotalBytes: 16 << 30, AvailableBytes: 8 << 30, CompressedBytes: 2 << 30}, want: PressureWarning},
		{name: "critical compressor", mem: Memory{TotalBytes: 16 << 30, AvailableBytes: 8 << 30, CompressedBytes: 4 << 30}, want: PressureCritical},
		{name: "critical wired", mem: Memory{TotalBytes: 16 << 30, AvailableBytes: 8 << 30, WiredBytes: 7 << 30}, want: PressureCritical},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got := AdmissionSampleFor(tc.mem)
			if got.Pressure != tc.want {
				t.Fatalf("pressure=%q want %q sample=%+v", got.Pressure, tc.want, got)
			}
			if got.AllocatableBytes != tc.mem.AvailableBytes {
				t.Fatalf("allocatable=%d want %d", got.AllocatableBytes, tc.mem.AvailableBytes)
			}
		})
	}
}

// TestObservedDarwinPressureOverridesOccupancyFallback pins #12701: wired and
// compressed byte counts describe occupancy, while the Darwin memorystatus level
// reports current pressure. A recognized live level therefore takes precedence;
// the old conservative occupancy rule remains the fallback when that probe is absent.
func TestObservedDarwinPressureOverridesOccupancyFallback(t *testing.T) {
	// JSON setup keeps this regression source-compatible with the pre-fix Memory
	// shape: old code ignores the two new fields and exhibits the false refusal.
	var base Memory
	if err := json.Unmarshal([]byte(`{
		"TotalBytes":20000000000,
		"AvailableBytes":5000000000,
		"WiredBytes":9000000000,
		"CompressedBytes":4000000000,
		"Pressure":"normal",
		"PressureKnown":true
	}`), &base); err != nil {
		t.Fatal(err)
	}

	t.Run("normal pressure admits fitting load despite high occupancy", func(t *testing.T) {
		if got := AdmissionSampleFor(base); got.Pressure != PressureNormal {
			t.Fatalf("admission pressure=%q, want normal", got.Pressure)
		}
		snap := BuildSnapshot("darwin", base, nil)
		var fields map[string]any
		raw, err := json.Marshal(snap)
		if err != nil {
			t.Fatal(err)
		}
		if err := json.Unmarshal(raw, &fields); err != nil {
			t.Fatal(err)
		}
		if fields["pressure_observed"] != true || fields["pressure"] != string(PressureNormal) {
			t.Fatalf("snapshot pressure fields=%v, want normal/true", fields)
		}
		if got := Evaluate(snap, 4); got.Admit == nil || !*got.Admit {
			t.Fatalf("fitting load refused under observed normal pressure: %+v", got)
		}
	})

	t.Run("critical pressure refuses fitting load", func(t *testing.T) {
		mem := base
		if err := json.Unmarshal([]byte(`{
			"WiredBytes":100,
			"CompressedBytes":100,
			"Pressure":"critical"
		}`), &mem); err != nil {
			t.Fatal(err)
		}
		if got := AdmissionSampleFor(mem); got.Pressure != PressureCritical {
			t.Fatalf("admission pressure=%q, want critical", got.Pressure)
		}
		if got := Evaluate(BuildSnapshot("darwin", mem, nil), 4); got.Admit == nil || *got.Admit {
			t.Fatalf("critical pressure admitted load: %+v", got)
		}
	})

	t.Run("capacity shortfall still refuses under normal pressure", func(t *testing.T) {
		got := Evaluate(BuildSnapshot("darwin", base, nil), 6)
		if got.Admit == nil || *got.Admit || got.ShortfallGB != 1 {
			t.Fatalf("short load capacity gate=%+v, want one GB refusal", got)
		}
	})

	t.Run("missing pressure probe preserves occupancy fallback", func(t *testing.T) {
		mem := base
		if err := json.Unmarshal([]byte(`{"PressureKnown":false}`), &mem); err != nil {
			t.Fatal(err)
		}
		if got := AdmissionSampleFor(mem); got.Pressure != PressureCritical {
			t.Fatalf("fallback admission pressure=%q, want critical", got.Pressure)
		}
		if got := Evaluate(BuildSnapshot("darwin", mem, nil), 4); got.Admit == nil || *got.Admit {
			t.Fatalf("missing-probe occupancy fallback admitted load: %+v", got)
		}
	})
}
